db_stress_tool/expected_state.cc

```diff
@@ -443,13 +475,56 @@ class ExpectedStateTraceRecordHandler : public TraceRecord::Handler,
                  const Slice& value) override {
     Slice key =
         StripTimestampFromUserKey(key_with_ts, FLAGS_user_timestamp_size);
+
+    bool should_buffer_write = (buffered_writes_ != nullptr);
+    if (should_buffer_write) {
+      return WriteBatchInternal::Merge(buffered_writes_, column_family_id, key,
+                                       value);
+    }
+
     return PutCF(column_family_id, key, value);
   }
 
+  Status MarkBeginPrepare(bool = false) override {
+    assert(buffered_writes_ == nullptr);
+
+    buffered_writes_ = new WriteBatch();
+    return Status::OK();
+  }
+
+  Status MarkEndPrepare(const Slice& xid) override {
+    assert(buffered_writes_ != nullptr);
+    std::string xid_str = xid.ToString();
+    assert(xid_to_buffered_writes_.find(xid_str) ==
+           xid_to_buffered_writes_.end());
+
+    xid_to_buffered_writes_[xid_str] = buffered_writes_;
+
+    buffered_writes_ = nullptr;
+
+    return Status::OK();
+  }
+
+  Status MarkCommit(const Slice& xid) override {
+    std::string xid_str = xid.ToString();
+    assert(xid_to_buffered_writes_.find(xid_str) !=
+           xid_to_buffered_writes_.end());
+    assert(xid_to_buffered_writes_.at(xid_str) != nullptr);
+
+    Status s = xid_to_buffered_writes_.at(xid_str)->Iterate(this);
+
+    delete xid_to_buffered_writes_.at(xid_str);
+    xid_to_buffered_writes_.erase(xid_str);
+
+    return s;
+  }
+
  private:
   uint64_t num_write_ops_ = 0;
   uint64_t max_write_ops_;
   ExpectedState* state_;
+  std::map<std::string, WriteBatch*> xid_to_buffered_writes_;
```

Comment:
Does it have to be ordered?

Reply:
Can we use `unique_ptr` for buffered writes, because we are sequentially processing the trace file, thus no concurrent (also shared) access/ownership.

Reply:
Fixed to unordered map. Fixed the raw pointer along with some != issues with unique_ptr.